File: app/api/schemas/document_versions.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class DocumentVersionResponse(BaseModel):
    id: str
    size: int
    created_at: float
    label: str
    source: str


class DocumentHistoryResponse(BaseModel):
    path: str
    head: str | None
    can_undo: bool
    can_redo: bool
    versions: list[DocumentVersionResponse]


class DocumentVersionActionRequest(BaseModel):
    path: str = Field(min_length=1, max_length=4096)
    action: Literal["checkpoint", "restore", "undo", "redo"]
    version_id: str | None = Field(default=None, max_length=64)
    label: str = Field(default="", max_length=2000)

File: web/src/api/client/documentVersions.ts
import { apiUrl } from '../base-url'
import { parseDetailOrThrow } from './_shared'

export interface DocumentVersion {
  id: string
  size: number
  /** Seconds since the epoch. */
  created_at: number
  label: string
  source: 'baseline' | 'checkpoint' | 'change' | string
}

export interface DocumentHistory {
  path: string
  head: string | null
  can_undo: boolean
  can_redo: boolean
  versions: DocumentVersion[]
}

export type DocumentVersionAction =
  | { action: 'checkpoint'; label?: string }
  | { action: 'restore'; version_id: string }
  | { action: 'undo' }
  | { action: 'redo' }

function versionsUrl(sessionId: string): string {
  return apiUrl(`/team/${encodeURIComponent(sessionId)}/document-versions`)
}

/** A workspace document's versions; also records a change made unwatched. */
export async function getDocumentVersions(sessionId: string, path: string): Promise<DocumentHistory> {
  const res = await fetch(`${versionsUrl(sessionId)}?${new URLSearchParams({ path })}`, {
    headers: { Accept: 'application/json' },
  })
  if (!res.ok) await parseDetailOrThrow(res, 'getDocumentVersions')
  return res.json()
}

/** Checkpoint, restore, undo or redo; the workspace file changes accordingly. */
export async function changeDocumentVersion(
  sessionId: string,
  path: string,
  change: DocumentVersionAction,
): Promise<DocumentHistory> {
  const res = await fetch(versionsUrl(sessionId), {
    method: 'POST',
    headers: { Accept: 'application/json', 'Content-Type': 'application/json' },
    body: JSON.stringify({ path, ...change }),
  })
  if (!res.ok) await parseDetailOrThrow(res, 'changeDocumentVersion')
  return res.json()
}
